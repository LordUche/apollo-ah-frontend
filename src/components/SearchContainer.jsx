import React, { Component } from 'react';
import PropTypes from 'prop-types';
import formDataJSON from 'formdata-json';
import queryString from 'query-string';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { getSearchResult } from '../selectors/searchSelector';
import { search } from '../actions/searchAction';
import Search from '../views/Search';

/**
 * Container component for the search view
 * @export
 * @class SearchContainer
 * @extends {Component}
 */
class SearchContainer extends Component {
  state ={
    q: '',
    categoryId: '0',
  }

  async componentDidMount() {
    const { location } = this.props;
    const query = queryString.parse(location.search);
    await this.initValues(query);
    this.search(query);
  }

  initValues = (query) => {
    const { q, categoryId } = query;
    this.setState({
      q,
      categoryId,
    });
  }

  handleSearch = (event) => {
    event.preventDefault();
    const query = formDataJSON(new FormData(event.target));
    if (query.categoryId === '0') {
      delete query.categoryId;
    }
    this.search(query);
  }

  search = (query) => {
    const { search: performSearch } = this.props;
    performSearch(query);
  }

  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  render() {
    const { searchResult } = this.props;
    const { q, categoryId } = this.state;
    return (
      <Search
      result={searchResult}
      handleSearch={this.handleSearch}
      handleChange={this.handleChange}
      q={q}
      categoryId={categoryId}
      />
    );
  }
}

SearchContainer.propTypes = {
  searchResult: PropTypes.object.isRequired,
  search: PropTypes.func.isRequired,
  location: PropTypes.object.isRequired
};

const mapStateToProps = createStructuredSelector(
  {
    searchResult: getSearchResult,
  }
);
const mapDispatchToProps = dispatch => ({
  search: query => dispatch(search(query)),
});
export default connect(mapStateToProps, mapDispatchToProps)(SearchContainer);
