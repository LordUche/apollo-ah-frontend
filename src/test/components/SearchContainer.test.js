import React from 'react';
import SearchContainer from '../../components/SearchContainer';
import setup from '../setup';


describe('<SearchContainer>', () => {
  it('should render', () => {
    const wrapper = setup(<SearchContainer location={{}} />);
    expect(wrapper).toMatchSnapshot();
  });
});
