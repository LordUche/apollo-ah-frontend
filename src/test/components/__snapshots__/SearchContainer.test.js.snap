// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchContainer> should render 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(SearchContainer)
        location={Object {}}
      >
        <SearchContainer
          location={Object {}}
          search={[Function]}
          searchResult={
            Object {
              "articles": Array [],
              "loading": false,
              "message": null,
              "page": Object {},
            }
          }
        >
          <Search
            categoryId="0"
            handleChange={[Function]}
            handleSearch={[Function]}
            q=""
            result={
              Object {
                "articles": Array [],
                "loading": false,
                "message": null,
                "page": Object {},
              }
            }
          >
            <main
              className="main-body"
            >
              <section
                className="search-cover"
              >
                <div
                  className="container"
                >
                  <form
                    className="search"
                    noValidate={true}
                    onSubmit={[Function]}
                  >
                    <div
                      className="row"
                    >
                      <div
                        className="col-sm-6 col-8"
                      >
                        <i
                          className="fas fa-search search-icon"
                        />
                        <input
                          className="search-input no-focus"
                          name="q"
                          onChange={[Function]}
                          value=""
                        />
                      </div>
                      <div
                        className="col-sm-3 d-none d-sm-inline-block"
                      >
                        <select
                          className="custom-select search-select no-focus"
                          name="categoryId"
                          onChange={[Function]}
                          value="0"
                        >
                          <option
                            value={0}
                          >
                            All categories
                          </option>
                          <option
                            value={3}
                          >
                            Business
                          </option>
                          <option
                            value={7}
                          >
                            Entertainment
                          </option>
                          <option
                            value={4}
                          >
                            Fashion
                          </option>
                          <option
                            value={6}
                          >
                            Food
                          </option>
                          <option
                            value={5}
                          >
                            Health
                          </option>
                          <option
                            value={8}
                          >
                            History
                          </option>
                          <option
                            value={1}
                          >
                            Technology
                          </option>
                          <option
                            value={2}
                          >
                            Sports
                          </option>
                          <option
                            value={9}
                          >
                            Others
                          </option>
                        </select>
                      </div>
                      <div
                        className="text-right col-sm-3 col-4"
                      >
                        <button
                          className="btn btn-brand search-button shadow-none"
                          type="submit"
                        >
                          Search
                        </button>
                      </div>
                    </div>
                  </form>
                </div>
              </section>
              <section
                className="wrapper min-vh-100 "
              >
                <div
                  className="container"
                >
                  <header
                    className="page-header page-header-sm text-center"
                  >
                    <h4>
                       No articles match your search result...
                    </h4>
                  </header>
                </div>
              </section>
            </main>
          </Search>
        </SearchContainer>
      </Connect(SearchContainer)>
    </Router>
  </BrowserRouter>
</Provider>
`;
